File: dexfile/src/main/java/com/github/netomi/bat/dexfile/visitor/DataItemVisitor.java
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile.visitor;

import com.github.netomi.bat.dexfile.*;
import com.github.netomi.bat.dexfile.annotation.*;
import com.github.netomi.bat.dexfile.debug.DebugInfo;

/**
 * @author Thomas Neidhart
 */
public interface DataItemVisitor {
    default void visitAnyDataItem(DexFile dexFile, DataItem dataItem) {
        throw new RuntimeException("Need to implement in class '" + this.getClass().getName() + "'.");
    }

    default void visitHeader(DexFile dexFile, DexHeader header) {
        visitAnyDataItem(dexFile, header);
    }

    default void visitMapList(DexFile dexFile, MapList mapList) {
        visitAnyDataItem(dexFile, mapList);
    }

    default void visitStringID(DexFile dexFile, StringID stringID) {
        visitAnyDataItem(dexFile, stringID);
    }

    default void visitStringData(DexFile dexFile, StringID stringID, StringData stringData) {
        visitAnyDataItem(dexFile, stringData);
    }

    default void visitTypeID(DexFile dexFile, TypeID typeID) {
        visitAnyDataItem(dexFile, typeID);
    }

    default void visitProtoID(DexFile dexFile, ProtoID protoID) {
        visitAnyDataItem(dexFile, protoID);
    }

    default void visitParameterTypes(DexFile dexFile, ProtoID protoID, TypeList typeList) {
        visitAnyDataItem(dexFile, typeList);
    }

    default void visitFieldID(DexFile dexFile, FieldID fieldID) {
        visitAnyDataItem(dexFile, fieldID);
    }

    default void visitMethodID(DexFile dexFile, MethodID methodID) {
        visitAnyDataItem(dexFile, methodID);
    }

    default void visitClassDef(DexFile dexFile, ClassDef classDef) {
        visitAnyDataItem(dexFile, classDef);
    }

    default void visitClassData(DexFile dexFile, ClassDef classDef, ClassData classData) {
        visitAnyDataItem(dexFile, classData);
    }

    default void visitInterfaceTypes(DexFile dexFile, ClassDef classDef, TypeList typeList) {
        visitAnyDataItem(dexFile, typeList);
    }

    default void visitStaticValuesArray(DexFile dexFile, ClassDef classDef, EncodedArray encodedArray) {
        visitAnyDataItem(dexFile, encodedArray);
    }

    default void visitAnnotationsDirectory(DexFile dexFile, ClassDef classDef, AnnotationsDirectory annotationsDirectory) {
        visitAnyDataItem(dexFile, annotationsDirectory);
    }

    default void visitClassAnnotations(DexFile dexFile, AnnotationsDirectory annotationsDirectory, AnnotationSet annotationSet) {
        visitAnyDataItem(dexFile, annotationSet);
    }

    default void visitFieldAnnotations(DexFile dexFile, FieldAnnotation fieldAnnotation, AnnotationSet annotationSet) {
        visitAnyDataItem(dexFile, annotationSet);
    }

    default void visitMethodAnnotations(DexFile dexFile, MethodAnnotation methodAnnotation, AnnotationSet annotationSet) {
        visitAnyDataItem(dexFile, annotationSet);
    }

    default void visitParameterAnnotations(DexFile dexFile, ParameterAnnotation parameterAnnotation, AnnotationSetRefList annotationSetRefList) {
        visitAnyDataItem(dexFile, annotationSetRefList);
    }

    default void visitAnnotationSet(DexFile dexFile, AnnotationSetRef annotationSetRef, AnnotationSet annotationSet) {
        visitAnyDataItem(dexFile, annotationSet);
    }

    default void visitAnnotation(DexFile dexFile, AnnotationSet annotationSet, int index, Annotation annotation) {
        visitAnyDataItem(dexFile, annotation);
    }

    default void visitCode(DexFile dexFile, EncodedMethod encodedMethod, Code code) {
        visitAnyDataItem(dexFile, code);
    }

    default void visitDebugInfo(DexFile dexFile, Code code, DebugInfo debugInfo) {
        visitAnyDataItem(dexFile, debugInfo);
    }

    default void visitCallSiteID(DexFile dexFile, CallSiteID callSiteID) {
        visitAnyDataItem(dexFile, callSiteID);
    }

    default void visitCallSite(DexFile dexFile, CallSiteID callSiteID, CallSite callSite) {
        visitAnyDataItem(dexFile, callSite);
    }

    default void visitMethodHandle(DexFile dexFile, MethodHandle methodHandle) {
        visitAnyDataItem(dexFile, methodHandle);
    }
}

File: dexfile/src/main/java/com/github/netomi/bat/dexfile/visitor/DataItemVisitor.kt
/*
 *  Copyright (c) 2020 Thomas Neidhart.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.github.netomi.bat.dexfile.visitor

import com.github.netomi.bat.dexfile.*
import com.github.netomi.bat.dexfile.annotation.*
import com.github.netomi.bat.dexfile.annotation.Annotation
import com.github.netomi.bat.dexfile.debug.DebugInfo

internal interface DataItemVisitor {
    fun visitAnyDataItem(dexFile: DexFile, dataItem: DataItem) {
        throw RuntimeException("Need to implement in class '${this.javaClass.name}'.")
    }

    fun visitHeader(dexFile: DexFile, header: DexHeader) {
        visitAnyDataItem(dexFile, header)
    }

    fun visitMapList(dexFile: DexFile, mapList: MapList) {
        visitAnyDataItem(dexFile, mapList)
    }

    fun visitStringID(dexFile: DexFile, stringID: StringID) {
        visitAnyDataItem(dexFile, stringID)
    }

    fun visitStringData(dexFile: DexFile, stringID: StringID, stringData: StringData) {
        visitAnyDataItem(dexFile, stringData)
    }

    fun visitTypeID(dexFile: DexFile, typeID: TypeID) {
        visitAnyDataItem(dexFile, typeID)
    }

    fun visitProtoID(dexFile: DexFile, protoID: ProtoID) {
        visitAnyDataItem(dexFile, protoID)
    }

    fun visitParameterTypes(dexFile: DexFile, protoID: ProtoID, typeList: TypeList) {
        visitAnyDataItem(dexFile, typeList)
    }

    fun visitFieldID(dexFile: DexFile, fieldID: FieldID) {
        visitAnyDataItem(dexFile, fieldID)
    }

    fun visitMethodID(dexFile: DexFile, methodID: MethodID) {
        visitAnyDataItem(dexFile, methodID)
    }

    fun visitClassDef(dexFile: DexFile, classDef: ClassDef) {
        visitAnyDataItem(dexFile, classDef)
    }

    fun visitClassData(dexFile: DexFile, classDef: ClassDef, classData: ClassData) {
        visitAnyDataItem(dexFile, classData)
    }

    fun visitInterfaceTypes(dexFile: DexFile, classDef: ClassDef, typeList: TypeList) {
        visitAnyDataItem(dexFile, typeList)
    }

    fun visitStaticValuesArray(dexFile: DexFile, classDef: ClassDef, encodedArray: EncodedArray) {
        visitAnyDataItem(dexFile, encodedArray)
    }

    fun visitAnnotationsDirectory(dexFile: DexFile, classDef: ClassDef, annotationsDirectory: AnnotationsDirectory) {
        visitAnyDataItem(dexFile, annotationsDirectory)
    }

    fun visitAnyAnnotationSet(dexFile: DexFile, annotationSet: AnnotationSet) {
        visitAnyDataItem(dexFile, annotationSet)
    }

    fun visitClassAnnotations(dexFile: DexFile, annotationsDirectory: AnnotationsDirectory, annotationSet: AnnotationSet) {
        visitAnyAnnotationSet(dexFile, annotationSet)
    }

    fun visitFieldAnnotations(dexFile: DexFile, fieldAnnotation: FieldAnnotation, annotationSet: AnnotationSet) {
        visitAnyAnnotationSet(dexFile, annotationSet)
    }

    fun visitMethodAnnotations(dexFile: DexFile, methodAnnotation: MethodAnnotation, annotationSet: AnnotationSet) {
        visitAnyAnnotationSet(dexFile, annotationSet)
    }

    fun visitParameterAnnotations(dexFile: DexFile, parameterAnnotation: ParameterAnnotation, annotationSetRefList: AnnotationSetRefList) {
        visitAnyDataItem(dexFile, annotationSetRefList)
    }

    fun visitAnnotationSet(dexFile: DexFile, annotationSetRef: AnnotationSetRef, annotationSet: AnnotationSet) {
        visitAnyAnnotationSet(dexFile, annotationSet)
    }

    fun visitAnnotation(dexFile: DexFile, annotationSet: AnnotationSet, index: Int, annotation: Annotation) {
        visitAnyDataItem(dexFile, annotation)
    }

    fun visitCode(dexFile: DexFile, encodedMethod: EncodedMethod, code: Code) {
        visitAnyDataItem(dexFile, code)
    }

    fun visitDebugInfo(dexFile: DexFile, code: Code, debugInfo: DebugInfo) {
        visitAnyDataItem(dexFile, debugInfo)
    }

    fun visitCallSiteID(dexFile: DexFile, callSiteID: CallSiteID) {
        visitAnyDataItem(dexFile, callSiteID)
    }

    fun visitCallSite(dexFile: DexFile, callSiteID: CallSiteID, callSite: CallSite) {
        visitAnyDataItem(dexFile, callSite)
    }

    fun visitMethodHandle(dexFile: DexFile, methodHandle: MethodHandle) {
        visitAnyDataItem(dexFile, methodHandle)
    }
}